File: code/C++/Chapter13/13-4/kmp.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

static vector<int> get_nexts(const string p) {
    vector<int> next = vector<int>(p.size(), 0);
    int j = 0;

    for (int i = 2; i < (int)p.size(); i++) {
        while (j != 0 && p[j] != p[i - 1]) {
            j = next[j];
        }
        if (p[j] == p[i - 1]) {
            j++;
        }
        next[i] = j;
    }
    return next;
}

int kmp(const string s, const string p) {
    vector<int> next = get_nexts(p);
    int j = 0;

    for (int i = 0; i < (int)s.size(); i++) {
        while (j > 0 && s[i] != p[j]) {
            j = next[j];
        }
        if (s[i] == p[j]) {
            j++;
        }
        if (j == (int)p.size()) {
            return i - p.size() + 1;
        }
    }

    return -1;
}

int main() {
    string s = "GTGTGAGCTGGTGTGTGCFAA";
    string p = "GTGTGCF";
    cout << kmp(s, p) << endl;
    return 0;
}

File: code/Java/Chapter13/13-4/KMP.java
public class KMP {
    public static void main(String[] args) {
        String s = "GTGTGAGCTGGTGTGTGCFAA";
        String p = "GTGTGCF";
        System.out.println(kmp(s, p));
    }

    public static int[] getNexts(String p) {
        int[] next = new int[p.length()];
        int j = 0;

        for (int i = 2; i < p.length(); i++) {
            while (j != 0 && p.charAt(j) != p.charAt(i - 1)) {
                j = next[j];
            }
            if (p.charAt(j) == p.charAt(i - 1)) {
                j++;
            }
            next[i] = j;
        }
        return next;
    }

    public static int kmp(String s, String p) {
        int[] next = getNexts(p);
        int j = 0;

        for (int i = 0; i < s.length(); i++) {
            while (j > 0 && s.charAt(i) != p.charAt(j)) {
                j = next[j];
            }
            if (s.charAt(i) == p.charAt(j)) {
                j++;
            }
            if (j == p.length()) {
                return i - p.length() + 1;
            }
        }
        return -1;
    }
}
